### DISCUSSION TITLE
Gauging interest/compatibility for WebGPU backend/Request

### DISCUSSION THREAD

**Discussion by jon-chuang** [intel/llvm#5614] [Ideas]
Apparently, there should exist a path from SYCL source to WebGPU backend, by writing an abstraction layer that takes SYCL calls and converts them to WebGPU C API calls (see [here](https://github.com/emscripten-core/emscripten/issues/9575#issuecomment-585941443) and the [header file](https://github.com/gfx-rs/wgpu/blob/master/ffi/wgpu.h)). WebGPU exposes graphics interface, but also allows general compute and should allow for Vulkan compatible SPIR-V code, which can be converted from OpenCL SPIR-V via clspv.

Hence, I was wondering what the process might be to compile these Vulkan SPIR-V bytecode files through sycl-llvm SPIR-V codegen and pass through clspv, and link with a SYCL runtime that makes wgpu.h calls? 

I understand that SYCL is primarily of interest to Intel in the HPC/data center space to program their complex suite of devices through a single API. However, having compatibility with a web-facing standard that works flawlessly could be in Intel's and SYCL's favour. This is because it could encourage adoption of SYCL by those looking to maintain a single code-base for both their web users and for data center scale accelerations, for instance if they themselves have customers with varying compute requirements.

**Comment by jon-chuang** [comment]
Can I get some response on this issue?

**Comment by bader** [comment]
> WebGPU exposes graphics interface, but also allows general compute and should allow for Vulkan compatible SPIR-V code, which can be converted from OpenCL SPIR-V via clspv.

AFAIK, clspv compiles "OpenCL C" source code to SPIR-V shaders at the cost of limiting OpenCL features. So I don't think proposed compilation flow is possible with the available tools.

Do you have any particular examples where using SYCL programming model for WebGPU back-end is beneficial?

**Comment by matkara** [comment]
1. naive example: a way to enable SYCL programming model for WebAssembly targets?

2. a way to target other technologies / OS by delegating the WebGPU calls to [Dawn](https://dawn.googlesource.com/dawn)?
Dawn translates the WebGPU calls to a natively supported technology (this is used by web browsers such as Chromium).
Dawn supports D3D12 shaders, Metal, Vulkan, OpenGL.
The Metal support could help with : https://github.com/intel/llvm/issues/982
